internal/cmd/restorer.go

```diff
@@ -0,0 +1,350 @@
+		r.totalRetries++
+		numLoaded, retries, err = r.writeBatchesWithRetry(ctx, batchesToBeCommitted)
+		if err != nil {
+			return fmt.Errorf("failed to write retried batch: %w", err)
+		}
+
+		retries++ // account for the initial attempt
+		r.writtenBatches += int64(len(batchesToBeCommitted))
+		r.writtenRels += int64(numLoaded)
+	case conflict && (!r.touchOnConflicts && !r.skipOnConflicts):
+		r.bar.Describe("conflict detected, aborting restore")
+		return fmt.Errorf("duplicate relationships found")
+	case retryable:
+		r.bar.Describe("retrying after error")
+		r.totalRetries++
+		numLoaded, retries, err = r.writeBatchesWithRetry(ctx, batchesToBeCommitted)
+		if err != nil {
+			return fmt.Errorf("failed to write retried batch: %w", err)
+		}
+
+		retries++ // account for the initial attempt
+		r.writtenBatches += int64(len(batchesToBeCommitted))
+		r.writtenRels += int64(numLoaded)
+	default:
+		r.bar.Describe("restoring from backup")
+		r.writtenBatches += int64(len(batchesToBeCommitted))
+	}
+
+	// it was a successful transaction commit without duplicates
+	if resp != nil {
+		r.writtenRels += int64(resp.NumLoaded)
+		if expectedLoaded != resp.NumLoaded {
+			log.Warn().Uint64("loaded", resp.NumLoaded).Uint64("expected", expectedLoaded).Msg("unexpected number of relationships loaded")
+		}
+	}
+
+	if err := r.bar.Set64(r.writtenRels + r.skippedRels); err != nil {
+		return fmt.Errorf("error incrementing progress bar: %w", err)
+	}
+
+	if !isatty.IsTerminal(os.Stderr.Fd()) {
+		log.Trace().
+			Int64("batches_written", r.writtenBatches).
+			Int64("relationships_written", r.writtenRels).
+			Int64("duplicate_batches", r.duplicateBatches).
+			Int64("duplicate_relationships", r.duplicateRels).
+			Int64("skipped_batches", r.skippedBatches).
+			Int64("skipped_relationships", r.skippedRels).
+			Uint64("retries", retries).
+			Msg("restore progress")
+	}
+
+	return nil
+}
+
+// writeBatchesWithRetry writes a set of batches using touch semantics and without transactional guarantees -
+// each batch will be committed independently. If a batch fails, it will be retried up to 10 times with a backoff.
+func (r *restorer) writeBatchesWithRetry(ctx context.Context, batches [][]*v1.Relationship) (uint64, uint64, error) {
+	backoffInterval := backoff.NewExponentialBackOff()
+	backoffInterval.InitialInterval = 10 * time.Millisecond
+	backoffInterval.MaxInterval = 2 * time.Second
+	backoffInterval.MaxElapsedTime = 0
+	backoffInterval.Reset()
+
+	var currentRetries, totalRetries, loadedRels uint64
+	for _, batch := range batches {
+		updates := lo.Map[*v1.Relationship, *v1.RelationshipUpdate](batch, func(item *v1.Relationship, index int) *v1.RelationshipUpdate {
+			return &v1.RelationshipUpdate{
+				Relationship: item,
+				Operation:    v1.RelationshipUpdate_OPERATION_TOUCH,
+			}
+		})
+
+		for {
+			// throttle the writes so we don't overwhelm the server
+			time.Sleep(backoffInterval.NextBackOff())
```

Comment:
add a log here, perhaps changing the progress bar message temporarily?

Reply:
yeah better to change the progress bar - logs mixed with the progress bar do not render well.

Reply:
addressed